File: backend/src/models/Notification.ts
import mongoose, { Schema, type Document } from 'mongoose';

export type NotificationType = 'ASSIGNMENT' | 'VERIFICATION' | 'SYSTEM' | 'ALERT';

export interface INotification extends Document {
  userId: mongoose.Types.ObjectId;
  type: NotificationType;
  title: string;
  message: string;
  readAt?: Date;
  metadata?: Record<string, unknown>;
  channel: 'MOCK_FCM' | 'SYSTEM';
  createdAt: Date;
  updatedAt: Date;
}

const notificationSchema = new Schema<INotification>(
  {
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    type: {
      type: String,
      enum: ['ASSIGNMENT', 'VERIFICATION', 'SYSTEM', 'ALERT'],
      default: 'SYSTEM'
    },
    title: { type: String, required: true, trim: true },
    message: { type: String, required: true, trim: true },
    readAt: { type: Date },
    metadata: { type: Schema.Types.Mixed, default: {} },
    channel: {
      type: String,
      enum: ['MOCK_FCM', 'SYSTEM'],
      default: 'MOCK_FCM'
    }
  },
  { timestamps: true }
);

notificationSchema.index({ userId: 1, createdAt: -1 });
notificationSchema.index({ type: 1, createdAt: -1 });

export const Notification = mongoose.model<INotification>('Notification', notificationSchema);

File: backend/src/main/java/com/raava/model/Notification.java
package com.raava.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.Map;

@Document(collection = "notifications")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Notification {
    @Id
    private String id;

    @DBRef(lazy = true)
    private User user;

    private String title;
    private String message;
    private Type type = Type.SYSTEM;
    private String channel = "MOCK_FCM";
    private Map<String, Object> metadata;
    private Instant readAt;

    @Indexed
    private Instant createdAt;

    @Indexed
    private Instant updatedAt;

    public enum Type { ASSIGNMENT, VERIFICATION, SYSTEM, ALERT }
}
